File: Jan2021/02Jan/ValidateBinarySearchTree.cpp
/*
Problem name: Validate Binary Search Tree
Problem link: https://leetcode.com/explore/featured/card/top-interview-questions-easy/94/trees/625
*/

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool isValidBST(TreeNode* root) {
        return helper(root, (long long)INT_MAX+1, (long long)INT_MIN-1);
    }
    
    bool helper(TreeNode* root, long long minVal, long long maxVal){
        if(!root)   return true;
        if(root->val >= minVal || root->val <= maxVal)   return false;
        
        return helper(root->left, (long long)root->val, maxVal) && helper(root->right, minVal , (long long)root->val);        
    }
};

File: Jan2021/02Jan/ValidateBinarySearchTree.java
/*
Problem Name: 98. Validate Binary Search Tree
Problem Link: https://leetcode.com/problems/validate-binary-search-tree/
*/
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isValidBST(TreeNode root) {
        List<Integer> l = new ArrayList<Integer>();
        inorder(root, l);
        for(int i=1;i<l.size();i++){
            if(l.get(i) <= l.get(i-1))
                return false;
        }
        return true;
    }
    static void inorder(TreeNode root, List<Integer> l){
        if(root == null)
            return;
        inorder(root.left,l);
        l.add(root.val);
        inorder(root.right,l);
    }
}
